Extend README in several directions

adamroyjones opened the pull request:
Hello again—sorry about the delay following #244. I've had quite a lot on.

This PR contains a few documentation changes; the branch is split into individually examinable commits. The [penultimate commit](https://github.com/mvdan/gofumpt/commit/067cda6980b700840520a50d56832856b65cfe0e) in the branch contains a few references to the sensitivity of gofumpt to the version of Go used to build it. I've also added a TOC and shuffled a few bits and pieces around.

Please don't be shy about telling me to change less (or to change more)! This is your repository, after all.

If all's well and you like the direction, I'll try to—in a later branch—take a look at adding a bit of usage documentation (e.g. around `-lang`) and extending what appears in the Go reference.

mvdan commented on README.md:
```diff
@@ -440,13 +437,13 @@ type ZeroFields struct {
 
 Adjacent parameters with the same type should be grouped together
```
for consistency, you missed emphasizing this one

mvdan commented on README.md:
````diff
@@ -436,34 +457,42 @@ type ZeroFields struct {
+`gofumpt` is a replacement for `gofmt`, so you can simply `go install` it:
+
+```sh
+go install mvdan.cc/gofumpt@latest
````
you're moving the `go install` way down - IMHO this is a mistake as it is what most new users want quickly.

mvdan commented on README.md:
```diff
@@ -2,20 +2,44 @@
+- [Roadmap](#roadmap)
+- [Frequently Asked Questions](#frequently-asked-questions)
+- [Contributing](#contributing)
+- [License](#license)
```
I'm conflicted about an index. On one hand, it's nice to be able to jump around. On the other, the current README jumps straight to the information, whereas the new one fills nearly an entire page with just the index. Overall I think this is not a good idea - if the README is so long that it can't possibly be navigated directly, then we've done something wrong.

What probably got you in this direction is the fact that the "added rules" section is so long. I've wanted to make it vertically shorter for some time, such as by using an HTML table. I'd rather go in that direction.

mvdan reviewed (commented):
These changes are far larger than I expected :) they are hard to review as it is a significant refactor of the README.

See https://github.com/mvdan/gofumpt/issues/244#issuecomment-1281246025. I don't think we should document the current behavior as I'm more and more convinced that it is wrong.

If you can change this PR to just keep commit 1 with the minor fix, I'll merge. Commits 2 and 3 I'm not particularly excited about, as they hide the most commonly needed information further down.

adamroyjones commented:
I'm sorry for the excess. I tried to break the changes up into independently meaningful commits that would each be easy to read (or discard); I may well have failed.

---

Re: #244: I trust your judgement. If you think the maintenance burden is small enough to bear, then that's great to hear.

---

I've force-pushed a new version of the branch that contains two commits.

The first mirrors the first commit of the original PR, except I've also added your suggestion [here](https://github.com/mvdan/gofumpt/pull/248#discussion_r997343341).

The second contains a subset of the changes from the last commit of the original PR: it coalesces the two Vim-related subsections into one.

In particular, I've removed the index (or TOC) I added. You're right that the main intent was to provide a way to jump beyond the rules. It's always possible to have a TOC that only links to headers of a given maximal depth (e.g., h2), but I'll defer to you.

If you want, I can discard the second commit from this branch.

mvdan commented on README.md:
```diff
@@ -2,7 +2,7 @@
 [![Go Reference](https://pkg.go.dev/badge/mvdan.cc/gofumpt/format.svg)](https://pkg.go.dev/mvdan.cc/gofumpt/format)
 
-	go install mvdan.cc/gofumpt@latest
+    go install mvdan.cc/gofumpt@latest
```
please undo the indentation changes - tabs are fine

adamroyjones replied:
Sorry for that—I think it was a change automatically introduced by Prettier and that I didn't manually discard. I've now updated things.

mvdan reviewed (approved):
You did break the changes into commits, and that's appreciated :) I just didn't expect you to propose a refactor of the README structure, for example, and it seems somewhat unrelated.

These changes all look fine to me. The rules do take too much vertical space right now, but if I can't find any good alternative like inline HTML to make it use more horizontal space, then perhaps we could move them elsewhere like `RULES.md`. Either way, the README should be shorter.
